mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp

```diff
@@ -152,21 +184,27 @@ struct LLVMInlinerInterface : public DialectInlinerInterface {
       for (Attribute attr : *attrs) {
         auto attrDict = cast<DictionaryAttr>(attr);
         for (NamedAttribute attr : attrDict) {
-          if (attr.getName() == LLVM::LLVMDialect::getByValAttrName())
-            continue;
-          // TODO: Handle all argument attributes;
-          LLVM_DEBUG(llvm::dbgs() << "Cannot inline " << funcOp.getSymName()
-                                  << ": unhandled argument attribute \""
-                                  << attr.getName() << "\"\n");
-          return false;
+          if (disallowedArgumentAttrs.contains(attr.getName())) {
```

Comment:
A possible alternative may be to have a static function with a static and thread local member:
```
static bool isDisallowedArgumentAttr(NamedAttribute attr) {
  thread_local DenseSet<StringAttr> disallowedAttrs = {
    StringAttr::get(attr.getContext(), LLVM::LLVMDialect::getAlignAttrName()),
    ...
  };
  return disallowedAttrs.contains(attr.getName());
}
```
That would result in one copy of the set per thread if I understand correctly.

Reply:
Why `thread_local` and not just `static`?

Reply:
I assume the initialization of a static is not thread safe and it may happen in multiple inliner threads at the same time?

Reply:
`static` should be at load time, I think? Not really an expert here, to be honest

Reply:
Reporting back: This does not seem to work; the variables are indeed only initialized once, but `contains` never finds anything, probably because they have been initialized with StringAttrs from a different context, and thus the pointer comparison fails. So I think this would only work by doing string comparisons. Hence the current solution seems better. But I understand the controversy... Unfortunately it seems to me like there's no way of doing this without keeping state on the _specific_ dialect instantiation 🤕

Reply:
> static should be at load time, I think? Not really an expert here, to be honest

in C++ a static variable is initialized when the function is called for the first time AFAIK.

Reply:
> in C++ a static variable is initialized when the function is called for the first time AFAIK.

Yes, apparently this is the case in C++11 and onwards. It's thread safe, so that's not a problem. The problem is that StringAttrs are tried to a specific context, and the context can change throughout execution, so the lookup does not work when a new object is instantiated with a different context, because it will not re-initialize the maps.

The current solution with having the maps on the object itself is the "correct" one as far as I can tell 🙂